### DISCUSSION TITLE
Avoid Resteasy processing for custom HttpAuthenticationMechanism

### DISCUSSION THREAD

**Discussion by hillmanytli** [quarkusio/quarkus#26052] [Q&A]
I am implementing `HttpAuthenticationMechanism` for custom authentication, call it `CustomAuthenticationMechanism`. When the user hit GET `/api/impersonate?id={targetId}`, he will be authenticated as `targetId` and be redirected to a landing page. I have taken reference of what `FormAuthenticationMechanism` has done.

At the end of `CustomAuthenticationMechanism.authenticate()` method, it does followings
```
  exchange.response().setStatusCode(302);
  exchange.response().headers().add(HttpHeaderNames.LOCATION, location);
  exchange.response().end();
```
Basically this works. However in the server log, it raises following exception
```
ERROR [org.jbo.res.rea.com.cor.AbstractResteasyReactiveContext] (vert.x-eventloop-thread-1) Request failed: java.io.IOException: java.io.IOException: Failed to write
            at org.jboss.resteasy.reactive.server.vertx.ResteasyReactiveOutputStream.close(ResteasyReactiveOutputStream.java:272)
            at org.jboss.resteasy.reactive.server.jaxrs.WriterInterceptorContextImpl.proceed(WriterInterceptorContextImpl.java:61)
            at io.quarkus.smallrye.opentracing.runtime.QuarkusSmallRyeTracingStandaloneContainerResponseFilter.aroundWriteTo(QuarkusSmallRyeTracingStandaloneContainerResponseFilter.java:40)
...
```
It looks like after `CustomAuthenticationMechanism` has done, Resteasy is trying to handle the HTTP call again and fails when it writes to the response, which has already been written by `CustomAuthenticationMechanism`. What can I do to avoid this exception?

**Comment by sberyozkin** [comment]
`FormAuthenticationMechanism` redirects from its challenge method,  if you'd like to do it from `authenticate` to complete the flow then the option which works is to throw `AuthenticationRedirectException`

**Reply by hillmanytli** [reply]
In `FormAuthenticationMechanism` , there is an execution path that follows `authenticate` -> `runFormAuth` -> `handleRedirectBack` where it ends the response with 302, so I am trying to replicate this flow.

I will try throwing `AuthenticationRedirectException`. Will be happy to take this approach once verified.

**Reply by sberyozkin** [reply]
Sorry, missed that. I wonder if it the combination of trying to end the call in `authenticate` and RestEasy Reactive that causes the problem. Can you please try `FormAuthenticationMechanism` with Resteasy Reactive ? throwing `AuthenticationRedirectException` should do it, it won't be a very clean solution in a case like yours because this is a `positive` redirect  to a landing page...

**Reply by hillmanytli** [reply]
I will put together a simple project with `FormAuthenticationMechanism` and `CustomAuthenticationMechanism` to illustrate it, but got tied in these days. Will come back once it is ready.
